# Assignment #3: A simple language classifier

Author: Pierre Nugues

## Objectives

In this assignment, you will implement a language detector inspired from Google's _Compact language detector_, version 3 (CLD3): https://github.com/google/cld3. CLD3 is written in C++ and its code is available from GitHub. The objectives of the assignment are to:
* Write a program to classify languages
* Use neural networks
* Know what a classifier is
* Write a short report of 1 to 2 pages to describe your program. You will notably comment the performance you obtained and how you could improve it.

## Description

### System Overview

Read the GitHub description of CLD3, https://github.com/google/cld3, (_Model_ section). In your individual report you will:
1. Summarize the system in two or three sentences;
2. Outline the CLD3 overall architecture in a figure. Use building blocks only and do not specify the parameters.

## Imports

In [1]:
import bz2
import json
import os
import numpy as np
import requests
from sklearn.neural_network import MLPClassifier
from sklearn.feature_extraction import DictVectorizer
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score, classification_report
from sklearn.metrics import confusion_matrix

### Dataset

As dataset, we will use Tatoeba, https://tatoeba.org/eng/downloads. It consists of more than 8 million short texts in 347 languages and it is available in one file called `sentences.csv`.

The dataset is structured this way: There is one text per line, where each line consists of the three following fields separated by tabulations and ended by a carriage return:
```
sentence id [tab] language code [tab] text [cr]
```
Each text (sentence) has a unique id and has a language code that follows the ISO 639-3 standard (see below). 

### Scope of the lab

In this lab, you will consider three languages only: French (fra), English (eng), and Swedish (swe). Below is an excerpt of the Tatoeba dataset limited to these three languages: 

```
1276    eng     Let's try something.
1277    eng     I have to go to sleep.
1280    eng     Today is June 18th and it is Muiriel's birthday!
...
1115    fra     Lorsqu'il a demandé qui avait cassé la fenêtre, tous les garçons ont pris un air innocent.
1279    fra     Je ne supporte pas ce type.
1441    fra     Pour une fois dans ma vie je fais un bon geste... Et ça ne sert à rien.
...
337413  swe     Vi trodde att det var ett flygande tefat.
341910  swe     Detta är huset jag bodde i när jag var barn.
341938  swe     Vi hade roligt på stranden igår.
...
```
Tatoeba is updated continuously. The examples from this dataset come from a corpus your instructor downloaded on September 15, 2020.

### Understanding the $\mathbf{X}$ matrix (feature matrix)

You will now investigate the CLD3 features:
 *  What are the features CLD3 extracts from each text?
 * Create manually a simplified $\mathbf{X}$ matrix where you will represent the 9 texts with CLD3 features. You will use a restricted set of features: You will only consider the letters _a_, _b_, and _n_ and the bigrams _an_, _ba_, and _na_. You will ignore the the rest of letters and bigrams as well as the trigrams. Your matrix will have 9 rows and 6 columns, each column will contain these counts: `[#a, #b, #n, #an, #ba, #na]`.

The CLD3's original description uses relative frequencies (counts of a letter divided by the total counts of letters in the text). Here, you will use the raw counts. To help you, your instructor filled the fourth row of the matrix corresponding to the first text in French. Fill in the rest. You will include this matrix in your report. 

$\mathbf{X} =
\begin{bmatrix}
0& 0& 1& 0& 0& 0\\
1& 0& 0& 0& 0& 0\\
3& 1& 2& 1& 0& 0\\
8& 0& 8& 1& 0& 0\\
1& 0& 1& 0& 0& 0\\
4& 1& 6& 1& 0& 0\\
3& 0& 1& 1& 0& 0\\
5& 1& 2& 0& 0& 0\\
2& 0& 1& 1& 0& 0\\
\end{bmatrix}$
; $\mathbf{y} =
\begin{bmatrix}
     \text{eng} \\
     \text{eng}\\
     \text{eng}\\
    \text{fra}\\
   \text{fra}  \\
     \text{fra}\\
    \text{swe}\\
 \text{swe}   \\
 \text{swe}   
\end{bmatrix}$

## Programming: Extracting the features

Before you start programming, download the Tatoeba dataset.

### Loading and filtering the dataset

Run the code to read the dataset and split it into lines. You may have to change the path

In [2]:
dataset = open('corpus/sentences.csv', encoding='utf8').read().strip()
dataset = dataset.split('\n')
dataset[:10]

['1\tcmn\t我們試試看！',
 '2\tcmn\t我该去睡觉了。',
 '3\tcmn\t你在干什麼啊？',
 '4\tcmn\t這是什麼啊？',
 '5\tcmn\t今天是６月１８号，也是Muiriel的生日！',
 '6\tcmn\t生日快乐，Muiriel！',
 '7\tcmn\tMuiriel现在20岁了。',
 '8\tcmn\t密码是"Muiriel"。',
 '9\tcmn\t我很快就會回來。',
 '10\tcmn\t我不知道。']

Run the code to split the fields and remove possible whitespaces

In [3]:
dataset = list(map(lambda x: tuple(x.split('\t')), dataset))
dataset = list(map(lambda x: tuple(map(str.strip, x)), dataset))
dataset[:3]

[('1', 'cmn', '我們試試看！'), ('2', 'cmn', '我该去睡觉了。'), ('3', 'cmn', '你在干什麼啊？')]

Write the code to extract the French, English, and Swedish texts. You will call the resulting dataset: `dataset_small`

In [4]:
# Write your code here
correctCodes = ['eng', 'fra', 'swe']

dataset_small = list(filter(lambda x: x[1] in correctCodes, dataset))

In [5]:
dataset_small[:5]

[('1115',
  'fra',
  "Lorsqu'il a demandé qui avait cassé la fenêtre, tous les garçons ont pris un air innocent."),
 ('1276', 'eng', "Let's try something."),
 ('1277', 'eng', 'I have to go to sleep.'),
 ('1279', 'fra', 'Je ne supporte pas ce type.'),
 ('1280', 'eng', "Today is June 18th and it is Muiriel's birthday!")]

### Functions to Count Characters Ngrams

Write a function `count_chars(string, lc=True)` to count characters (unigrams) of a string. You will set the text in lowercase if `lc` is set to `True`. As in CLD3, you will return the relative frequencies of the unigrams.

In [6]:
# Write your code here
regex = "\p{WhiteSpace}"

def count_chars(string, lc=True):
    length = len(string)
    unigram_dic = {}
    
    if lc:
        string = string.lower()
    
    for unigram in string:
        if unigram in unigram_dic:
            unigram_dic[unigram] += 1 / length
        else:
            unigram_dic[unigram] = 1 / length
    
    return unigram_dic
    


Write a function `count_bigrams(string, lc=True)` to count the characters bigrams of a string. You will set the text in lowercase if `lc` is set to `True`. As in CLD3, you will return the relative frequencies of the bigrams.

In [7]:
# Write your code heredef count_chars(string, lc=True):
def count_bigrams(string, lc=True):
    length = len(string) - 1
    bigram_dic = {}
    
    if lc:
        string = string.lower()
    
    for index in range(length - 1):
        bigram = string[index] + string[index + 1]
        
        if bigram in bigram_dic:
            bigram_dic[bigram] += 1 / length
        else:
            bigram_dic[bigram] = 1 / length
    
    return bigram_dic


Write a function `count_trigrams(string, lc=True)` to count the characters trigrams of a string. You will set the text in lowercase if `lc` is set to `True`. As in CLD3, you will return the relative frequencies of the trigrams.

In [8]:
# Write your code here
def count_trigrams(string, lc=True):
    length = len(string) - 2
    trigram_dic = {}
    
    if lc:
        string = string.lower()
    
    for index in range(length - 2):
        trigram = string[index] + string[index + 1] + string[index + 2]
        
        if trigram in trigram_dic:
            trigram_dic[trigram] += 1 / length
        else:
            trigram_dic[trigram] = 1 / length
    
    return trigram_dic

In [9]:
count_chars("Let's try something.")

{'l': 0.05,
 'e': 0.1,
 't': 0.15000000000000002,
 "'": 0.05,
 's': 0.1,
 ' ': 0.1,
 'r': 0.05,
 'y': 0.05,
 'o': 0.05,
 'm': 0.05,
 'h': 0.05,
 'i': 0.05,
 'n': 0.05,
 'g': 0.05,
 '.': 0.05}

In [10]:
count_bigrams("Let's try something.")

{'le': 0.05263157894736842,
 'et': 0.10526315789473684,
 "t'": 0.05263157894736842,
 "'s": 0.05263157894736842,
 's ': 0.05263157894736842,
 ' t': 0.05263157894736842,
 'tr': 0.05263157894736842,
 'ry': 0.05263157894736842,
 'y ': 0.05263157894736842,
 ' s': 0.05263157894736842,
 'so': 0.05263157894736842,
 'om': 0.05263157894736842,
 'me': 0.05263157894736842,
 'th': 0.05263157894736842,
 'hi': 0.05263157894736842,
 'in': 0.05263157894736842,
 'ng': 0.05263157894736842}

In [11]:
count_trigrams("Let's try something.")

{'let': 0.05555555555555555,
 "et'": 0.05555555555555555,
 "t's": 0.05555555555555555,
 "'s ": 0.05555555555555555,
 's t': 0.05555555555555555,
 ' tr': 0.05555555555555555,
 'try': 0.05555555555555555,
 'ry ': 0.05555555555555555,
 'y s': 0.05555555555555555,
 ' so': 0.05555555555555555,
 'som': 0.05555555555555555,
 'ome': 0.05555555555555555,
 'met': 0.05555555555555555,
 'eth': 0.05555555555555555,
 'thi': 0.05555555555555555,
 'hin': 0.05555555555555555}

### Counting the ngrams in the dataset

You will now extract the features from each text. For this, add the character, bigram, and trigram relative frequencies to the texts using this format:
`(text_id, language_id, text, char_cnt, bigram_cnt, trigram_cnt)`.

From the datapoint:
`('1276', 'eng', "Let's try something.")`,
you must return:

`('1276', 'eng', "Let's try something.", 
  {'l': 0.05, 'e': 0.1, 't': 0.15, "'": 0.05, 's': 0.1, ' ': 0.1, 'r': 0.05, 'y': 0.05, 'o': 0.05, 'm': 0.05, 'h': 0.05, 'i': 0.05, 'n': 0.05, 'g': 0.05, '.': 0.05},
  {'le': 0.05263157894736842, 'et': 0.10526315789473684, "t'": 0.05263157894736842, "'s": 0.05263157894736842, 's ': 0.05263157894736842, ' t': 0.05263157894736842, 'tr': 0.05263157894736842, 'ry': 0.05263157894736842, 'y ': 0.05263157894736842, ' s': 0.05263157894736842, 'so': 0.05263157894736842, 'om': 0.05263157894736842, 'me': 0.05263157894736842, 'th': 0.05263157894736842, 'hi': 0.05263157894736842, 'in': 0.05263157894736842, 'ng': 0.05263157894736842, 'g.': 0.05263157894736842},
  {'let': 0.05555555555555555, "et'": 0.05555555555555555, "t's": 0.05555555555555555, "'s ": 0.05555555555555555, 's t': 0.05555555555555555, ' tr': 0.05555555555555555, 'try': 0.05555555555555555, 'ry ': 0.05555555555555555, 'y s': 0.05555555555555555, ' so': 0.05555555555555555, 'som': 0.05555555555555555, 'ome': 0.05555555555555555, 'met': 0.05555555555555555, 'eth': 0.05555555555555555, 'thi': 0.05555555555555555, 'hin': 0.05555555555555555, 'ing': 0.05555555555555555, 'ng.': 0.05555555555555555})`

You will store the extracted features in a list that you will call `dataset_small_feat`

In [12]:
# Write your code here
length = len(dataset_small)
dataset_small_feat = []

for index in range(length):
    tupel = ()
    data = dataset_small[index]
    
    line_nbr = data[0]
    language_id = data[1]
    text = data[2]
    
    unigram_dic = count_chars(text)
    bigram_dic = count_bigrams(text)
    trigram_dic = count_trigrams(text)
    
    tupel = (line_nbr, language_id, text, unigram_dic, bigram_dic, trigram_dic)
    
    dataset_small_feat.append(tupel)
        

In [13]:
dataset_small_feat[:2]

[('1115',
  'fra',
  "Lorsqu'il a demandé qui avait cassé la fenêtre, tous les garçons ont pris un air innocent.",
  {'l': 0.044444444444444446,
   'o': 0.05555555555555556,
   'r': 0.05555555555555556,
   's': 0.07777777777777778,
   'q': 0.022222222222222223,
   'u': 0.044444444444444446,
   "'": 0.011111111111111112,
   'i': 0.06666666666666667,
   ' ': 0.16666666666666663,
   'a': 0.08888888888888889,
   'd': 0.022222222222222223,
   'e': 0.05555555555555556,
   'm': 0.011111111111111112,
   'n': 0.08888888888888889,
   'é': 0.022222222222222223,
   'v': 0.011111111111111112,
   't': 0.05555555555555556,
   'c': 0.022222222222222223,
   'f': 0.011111111111111112,
   'ê': 0.011111111111111112,
   ',': 0.011111111111111112,
   'g': 0.011111111111111112,
   'ç': 0.011111111111111112,
   'p': 0.011111111111111112,
   '.': 0.011111111111111112},
  {'lo': 0.011235955056179775,
   'or': 0.011235955056179775,
   'rs': 0.011235955056179775,
   'sq': 0.011235955056179775,
   'qu': 0.02247191

The unigram frequencies

In [14]:
dataset_small_feat[0][3].items()

dict_items([('l', 0.044444444444444446), ('o', 0.05555555555555556), ('r', 0.05555555555555556), ('s', 0.07777777777777778), ('q', 0.022222222222222223), ('u', 0.044444444444444446), ("'", 0.011111111111111112), ('i', 0.06666666666666667), (' ', 0.16666666666666663), ('a', 0.08888888888888889), ('d', 0.022222222222222223), ('e', 0.05555555555555556), ('m', 0.011111111111111112), ('n', 0.08888888888888889), ('é', 0.022222222222222223), ('v', 0.011111111111111112), ('t', 0.05555555555555556), ('c', 0.022222222222222223), ('f', 0.011111111111111112), ('ê', 0.011111111111111112), (',', 0.011111111111111112), ('g', 0.011111111111111112), ('ç', 0.011111111111111112), ('p', 0.011111111111111112), ('.', 0.011111111111111112)])

The bigram frequencies

In [15]:
dataset_small_feat[0][4].items()

dict_items([('lo', 0.011235955056179775), ('or', 0.011235955056179775), ('rs', 0.011235955056179775), ('sq', 0.011235955056179775), ('qu', 0.02247191011235955), ("u'", 0.011235955056179775), ("'i", 0.011235955056179775), ('il', 0.011235955056179775), ('l ', 0.011235955056179775), (' a', 0.033707865168539325), ('a ', 0.02247191011235955), (' d', 0.011235955056179775), ('de', 0.011235955056179775), ('em', 0.011235955056179775), ('ma', 0.011235955056179775), ('an', 0.011235955056179775), ('nd', 0.011235955056179775), ('dé', 0.011235955056179775), ('é ', 0.02247191011235955), (' q', 0.011235955056179775), ('ui', 0.011235955056179775), ('i ', 0.011235955056179775), ('av', 0.011235955056179775), ('va', 0.011235955056179775), ('ai', 0.02247191011235955), ('it', 0.011235955056179775), ('t ', 0.02247191011235955), (' c', 0.011235955056179775), ('ca', 0.011235955056179775), ('as', 0.011235955056179775), ('ss', 0.011235955056179775), ('sé', 0.011235955056179775), (' l', 0.02247191011235955), ('la

## Programming: Building $\mathbf{X}$

You will now build the $\mathbf{X}$ matrix. In this assignment, you will only consider unigrams to speed up the training step. This means that you will set aside the character bigrams and trigrams.

When you are done with the lab requirements, feel free to improve the program and include bigrams and trigrams. To add bigrams, a possible method is to add the bigram dictionary to the unigram one using update and then to extract the resulting dictionary. You can easily extend this to trigrams. Feel free to use another method if you want.

In [16]:
INCLUDE_BIGRAMS = False
if INCLUDE_BIGRAMS:
    for i in range(len(dataset_small_feat)):
        dataset_small_feat[i][3].update(dataset_small_feat[i][4])

### Vectorizing the features

The CLD3 architecture uses embeddings. In this lab, we will simplify it and we will use a feature vector instead consisting of the character frequencies. For example, you will represent the text:

`"Let's try something."`

with:

`{'l': 0.05, 'e': 0.1, 't': 0.15, "'": 0.05, 's': 0.1, ' ': 0.1, 
 'r': 0.05, 'y': 0.05, 'o': 0.05, 'm': 0.05, 'h': 0.05, 'i': 0.05, 
 'n': 0.05, 'g': 0.05, '.': 0.05}`

To create the $\mathbf{X}$ matrix, we need to transform the dictionaries of `dataset_small` into numerical vectors. The `DictVectorizer` class from the scikit-learn library, see here [https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html], has two methods, `fit()` and `transform()`, and a combination of both `fit_transform()` to convert dictionaries into such vectors.

You will now write the code to:

1. Extract the character frequency dictionaries from `dataset_small` corresponding to its 3rd index and set them in a list;
2. Convert the list of dictionaries into an $\mathbf{X}$ matrix using `DictVectorizer`.

#### Extracting the character frequencies

Produce a new list of datapoints with the unigrams only. Each item in this list will be a dictionary. You will call it `X_cat`

In [17]:
# Write your code here

X_cat = []

for data in dataset_small:
    text = data[2]
    unigram_dic = count_chars(text)
    X_cat.append(unigram_dic)

In [18]:
X_cat[:2]

[{'l': 0.044444444444444446,
  'o': 0.05555555555555556,
  'r': 0.05555555555555556,
  's': 0.07777777777777778,
  'q': 0.022222222222222223,
  'u': 0.044444444444444446,
  "'": 0.011111111111111112,
  'i': 0.06666666666666667,
  ' ': 0.16666666666666663,
  'a': 0.08888888888888889,
  'd': 0.022222222222222223,
  'e': 0.05555555555555556,
  'm': 0.011111111111111112,
  'n': 0.08888888888888889,
  'é': 0.022222222222222223,
  'v': 0.011111111111111112,
  't': 0.05555555555555556,
  'c': 0.022222222222222223,
  'f': 0.011111111111111112,
  'ê': 0.011111111111111112,
  ',': 0.011111111111111112,
  'g': 0.011111111111111112,
  'ç': 0.011111111111111112,
  'p': 0.011111111111111112,
  '.': 0.011111111111111112},
 {'l': 0.05,
  'e': 0.1,
  't': 0.15000000000000002,
  "'": 0.05,
  's': 0.1,
  ' ': 0.1,
  'r': 0.05,
  'y': 0.05,
  'o': 0.05,
  'm': 0.05,
  'h': 0.05,
  'i': 0.05,
  'n': 0.05,
  'g': 0.05,
  '.': 0.05}]

#### Vectorize `X_cat`

Convert you `X_cat` matrix into a numerical representation using `DictVectorizer`: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html

In [64]:
# Write your code here
vec = DictVectorizer()
X = vec.fit_transform(X_cat)

## Programming: Building $\mathbf{y}$

You will now convert the list of language symbols into a $\mathbf{y}$ vector

Extract the language symbols from `dataset_small_feat` and call the resulting list `y_cat`

In [65]:
# Write your code here
y_cat = []

for data in dataset_small_feat:
    language_id = data[1]
    y_cat.append(language_id)

In [66]:
y_cat[:5]

['fra', 'eng', 'eng', 'fra', 'eng']

Extract the set of language symbols and build two indices mapping the symbols to integers and the integers to symbols. Both indices will be dictionaries that you will call: `lang2inx`and `inx2lang`.

In [67]:
# Write your code here
inx2lang = {0:'fra', 1:'eng', 2:'swe'}
lang2inx = {'fra':0, 'eng':1, 'swe':2}


In [68]:
inx2lang

{0: 'fra', 1: 'eng', 2: 'swe'}

In [69]:
lang2inx

{'fra': 0, 'eng': 1, 'swe': 2}

Convert your `y_cat` vector into a numerical vector. Call this vector `y`.

In [70]:
# Write your code here

y = list(map(lambda language_id: lang2inx[language_id], y_cat))

In [71]:
y[:5]

[0, 1, 1, 0, 1]

## Programming: Building the Model

Create a neural network using sklearn with a hidden layer of 50 nodes and a relu activation layer: https://scikit-learn.org/stable/modules/neural_networks_supervised.html. Set the maximal number of iterations to 5, in the beginning, and verbose to True. Use the default values for the rest. You will call you classifier `clf`

In [88]:
# Write your code here
clf = MLPClassifier(hidden_layer_sizes=(50,), max_iter = 10, verbose = True)

In [89]:
clf

MLPClassifier(hidden_layer_sizes=(50,), max_iter=10, verbose=True)

### Training and Validation Sets

You will now split the dataset into training and a validation sets

#### We shuffle the indices

In [90]:
indices = list(range(X.shape[0]))
np.random.shuffle(indices)
print(indices[:10])
X = X[indices, :]
y = np.array(y)[indices]

[283754, 54005, 197417, 99291, 145239, 120549, 63550, 11640, 238559, 59731]


#### We split the dataset

In [91]:
training_examples = int(X.shape[0] * 0.8)

X_train = X[:training_examples, :]
y_train = y[:training_examples]

X_val = X[training_examples:, :]
y_val = y[training_examples:]

### Fitting the model

Fit the model on the training set

In [92]:
# Write your code here
clf.fit(X_train, y_train)

Iteration 1, loss = 0.28809296
Iteration 2, loss = 0.10401515
Iteration 3, loss = 0.08243979
Iteration 4, loss = 0.07644067
Iteration 5, loss = 0.07379813
Iteration 6, loss = 0.07240124
Iteration 7, loss = 0.07133005
Iteration 8, loss = 0.07029194
Iteration 9, loss = 0.06943409
Iteration 10, loss = 0.06864319


/home/viktor/anaconda3/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(hidden_layer_sizes=(50,), max_iter=10, verbose=True)

## Predicting

Predict the `X_val` languages. You will call the result `y_val_pred`

In [93]:
# Write your code here
y_val_pred = clf.predict(X_val)

In [94]:
print(y_val_pred[:10])
print(y_val[:10])

[0 2 1 1 1 1 1 0 1 1]
[0 2 1 1 1 1 1 0 1 1]


#### Evaluating

In [95]:
# evaluate the model
y_symbols = ['fra', 'eng', 'swe']
accuracy_score(y_val, y_val_pred)

0.9741206689406599

In [96]:
print(classification_report(y_val, y_val_pred, target_names=y_symbols))
print('Micro F1:', f1_score(y_val, y_val_pred, average='micro'))
print('Macro F1', f1_score(y_val, y_val_pred, average='macro'))

              precision    recall  f1-score   support

         fra       0.96      0.97      0.97     26218
         eng       0.98      0.98      0.98     45550
         swe       0.97      0.87      0.92       645

    accuracy                           0.97     72413
   macro avg       0.97      0.94      0.95     72413
weighted avg       0.97      0.97      0.97     72413

Micro F1: 0.9741206689406599
Macro F1 0.9546554915024966


### Confusion Matrix

In [97]:
confusion_matrix(y_val, y_val_pred)

array([[25391,   823,     4],
       [  951, 44585,    14],
       [   13,    69,   563]])

Increase the number of iterations to improve the score. You may also change the parameters.

## Predict the language of a text

Now predict the languages of the strings below.

In [98]:
docs = ["Salut les gars !", "Hejsan grabbar!", "Hello guys!", "Hejsan tjejer!"]

Create features vectors from this list. Call this matrix `X_test`

In [99]:
# Write your code here
X_list = []

for sentence in docs:
    unigram_dic = count_chars(sentence)
    X_list.append(unigram_dic)
    
X_test = vec.transform(X_list)


And run the prediction that you will store in a variable called `pred_languages`

In [100]:
# Write your code here
pred_languages = clf.predict(X_test)

In [101]:
pred_languages = list(pred_languages)
pred_languages = list(map(lambda language_id: inx2lang[language_id], pred_languages))

## Submission

When you have written all the code and run all the cells, fill in your ID and as well as the name of the notebook.

In [102]:
STIL_ID = ["vi1146ol-s"] # Write your stil ids as a list
CURRENT_NOTEBOOK_PATH = os.path.join(os.getcwd(), 
                                     "33-language_detector.ipynb") # Write the name of your notebook

The submission code will send your answer. It consists of the predicted languages.

In [103]:
ANSWER = json.dumps({'pred_langs': pred_languages})
ANSWER

'{"pred_langs": ["fra", "swe", "eng", "swe"]}'

Now the moment of truth:
1. Save your notebook and
2. Run the cells below

In [104]:
SUBMISSION_NOTEBOOK_PATH = CURRENT_NOTEBOOK_PATH + ".submission.bz2"

In [105]:
ASSIGNMENT = 3
API_KEY = "f581ba347babfea0b8f2c74a3a6776a7"

# Copy and compress current notebook
with bz2.open(SUBMISSION_NOTEBOOK_PATH, mode="wb") as fout:
    with open(CURRENT_NOTEBOOK_PATH, "rb") as fin:
        fout.write(fin.read())

In [106]:
res = requests.post("https://vilde.cs.lth.se/edan20checker/submit", 
                    files={"notebook_file": open(SUBMISSION_NOTEBOOK_PATH, "rb")}, 
                    data={
                        "stil_id": STIL_ID,
                        "assignment": ASSIGNMENT,
                        "answer": ANSWER,
                        "api_key": API_KEY,
                    },
               verify=True)

# from IPython.display import display, JSON
res.json()

{'msg': None,
 'status': 'correct',
 'signature': '8a23364445b016854d730b2125f2f2118995003f4a2496dc94bc0d5663aa813c7ae9f1de157bc176b2212cf7cbc87789da52af1d52bc46e3131592d16cba2afe',
 'submission_id': 'd91aa600-7adb-42b7-a822-9eab65658a1a'}

## Postscript from Pierre Nugues

I created this assignment from an examination I wrote last year for the course on applied machine learning. I simplified it from the `README.md` on GitHub, https://github.com/google/cld3. I found the C++ code difficult to understand and I reimplemented a Keras/Tensorflow version of it from this `README`. Should you be interested, you can find it here: https://github.com/pnugues/language-detector.